### ISSUE TITLE
RF06 replaces function/procedure names with a segment type the grammar cannot match, so the fix is silently dropped

### ISSUE THREAD

**Issue by WAHIB-EL-KHADIRI** [sqlfluff/sqlfluff#8466]
Reporting this because the warning text asks for it, verbatim: *"Please report this as a bug with a minimal query which demonstrates this warning."*

## What happens

`sqlfluff fix` declines to apply an RF06 fix, saying the result would be unparsable. The result parses fine.

```sql
-- mysql
CREATE PROCEDURE `testprocedure`(in test int)
BEGIN
END~
```

```
$ sqlfluff fix --dialect mysql --rules RF06 procedure_in_param.sql
Fixes for RF06 not applied, as it would result in an unparsable file.
Please report this as a bug with a minimal query which demonstrates this warning.
```

The file is left unchanged, and the violation stays:

```
L: 1 | P: 18 | RF06 | Unnecessary quoted identifier `testprocedure`.
```

## The output would not have been unparsable

Both forms parse cleanly through the same API path, on `main` (`85dfbdff`):

```python
from sqlfluff.core import Linter, FluffConfig
L = Linter(config=FluffConfig(overrides={"dialect": "mysql"}))

for sql in ["CREATE PROCEDURE `testprocedure`(in test int)\nBEGIN\nEND~\n",
            "CREATE PROCEDURE  testprocedure (in test int)\nBEGIN\nEND~\n"]:
    p = L.parse_string(sql)
    print(len([v for v in p.violations if v.rule_code().startswith(("PRS", "LXR"))]),
          len(list(p.tree.recursive_crawl("unparsable"))))
```

```
0 0      <- quoted
0 0      <- unquoted
```

Same result for `mariadb` with `CALL \`somefunction\`('a');`.

So the guard is rejecting a fix it did not need to reject, and RF06 silently under-fixes as a result.

## Scope

Round-tripped the dialect fixture corpus on `main` and counted where the guard fires. Every hit is RF06, and every hit is a procedure or function identifier context:

| dialect | fixtures | examples |
|---|---|---|
| mysql | 9 | `procedure_in_param`, `procedure_out_param`, `procedure_definer`, `function_return`, `function_comment`, `call_statement`, `delimiter_procedure` |
| mariadb | 11 | `call_statement`, `create_or_replace_procedure`, `create_or_replace_function`, `function_definer`, `delimiter_procedure` |

Twenty of the project's own fixtures, each printing a "please report this as a bug" warning to any user who runs `fix` over them.

## Relationship to #8462

Same rule, opposite direction, and I think that is the interesting part.

- **#8462** — RF06 unquotes `'user'@'host'` account specifications and the output genuinely does not parse. The guard **does not catch it**, so corruption is written to disk.
- **This issue** — the guard **fires when it should not**, so a safe fix is skipped.

The guard is currently wrong in both directions on the same rule, which suggests the check itself is the thing to look at rather than RF06's individual cases.

Happy to open a PR if you have a preference for which way to take it. My guess, from #8415, is that the reachable common cause is what the guard re-parses rather than how RF06 builds the edit — but that is a guess, and I would rather hear yours before writing code against it.

## Disclosure

Found with a scanner I wrote that runs every fixture in `test/fixtures/dialects` through `fix` and re-parses the output; an AI coding agent helped write that tooling and this description. What I verified myself, from source rather than from a model: the two `parse_string` results above on a clean `main` checkout, that `sqlfluff fix --rules RF06` leaves both files byte-identical, and that all twenty hits report `RF06` and no other rule.

**Comment by WAHIB-EL-KHADIRI** [comment]
**Correcting my own framing here.** I titled this "the guard has false positives" and blamed the validator. I traced it properly and that is wrong: the validator is doing its job. The defect is in RF06. The symptom I reported is real, the attribution was not, and a precise cause is more useful than my guess — so here it is.

## Root cause

Turning on `sqlfluff.linter` debug for `mysql/procedure_in_param.sql` gives the failure chain:

```
Validation Check Fail for <FunctionNameSegment: ([L:1, P:18])>. Incomplete Match.
  Matched:   ()
  Unmatched: (<IdentifierSegment: ([L:1, P:18]) 'testprocedure'>,)
```

and it then cascades up through `CreateProcedureStatementSegment` -> `StatementSegment` -> `FileSegment`, which is why the whole fix is dropped.

The reason `FunctionNameSegment` matches nothing is a type mismatch, not a text problem. In `dialect_ansi.py`:

```python
FunctionNameIdentifierSegment=TypedParser("word", WordSegment, type="function_name_identifier")
```

and the slot is:

```python
OneOf(
    Ref("FunctionNameIdentifierSegment"),   # TypedParser on raw type "word"
    Ref("QuotedIdentifierSegment"),
)
```

RF06 builds its replacement from the dialect's `NakedIdentifierSegment`, which is a `RegexParser`:

```python
NakedIdentifierSegment(raw=identifier_contents, **naked_identifier_parser.segment_kwargs())
```

So after the fix the slot is offered a `naked_identifier`, and both branches of the `OneOf` fail — the `TypedParser` because the raw type is not `word`, and `QuotedIdentifierSegment` because the quotes were just removed. `Matched: ()`.

**The rule replaces a function-name identifier with a segment of a type that position's grammar cannot accept.** The resulting *text* is fine — that is why a full lex-and-parse of the output succeeds, as in my first comment — but the tree genuinely cannot be rebuilt, so the validator refusing is correct behaviour, not a false positive.

That also explains the shape of the 20 fixtures: every one is a procedure or function name, which is exactly the set of positions parsed by `TypedParser("word")` rather than by the general identifier grammar.

## Which leaves a real choice, and it is yours not mine

**A. Narrow — fix RF06.** Build the replacement with the parser appropriate to the slot rather than always reaching for `NakedIdentifierSegment`, or have the rule decline when the naked form cannot satisfy the position. Contained, and it does not touch anything else.

**B. Wider — make validation text-based.** `validate_segment_with_reparse` rematches existing segments (`self.raw_segments`) rather than re-lexing the fixed text. That is what makes it sensitive to a segment's *type* rather than to what the text would actually lex to.

I raise B only because it is the same mechanism behind #8415, where the type-based rematch is blind in the *other* direction: it cannot see two tokens fusing into one when whitespace is removed, so genuine corruption passes. One mechanism, opposite failure modes. But B is an architectural change to the fix loop and I am not going to pretend I know whether you want it.

Happy to write A if you point at which of the two shapes you prefer. I would rather ask than send you a PR built on a guess about your design.

## Relabelling

Retitling this from "the guard has false positives" to something accurate. #8462 is unaffected — RF06 corrupting `'user'@'host'` still reproduces on `main`, and the guard does not catch that one at all.

**Comment by WAHIB-EL-KHADIRI** [comment]
**Correcting the scope in my original report: "20 fixtures across mysql and mariadb" understates this, and I knew the number was only partially measured when I wrote it.**

That count came from scanning **two dialects**. The `GUARDED` detection did not exist when I ran the full corpus, so the other 26 dialects were never checked for it. Re-running with the detection on, **the guard fires outside MySQL and MariaDB**:

| dialect | fixtures |
|---|---|
| mariadb | 11 |
| mysql | 9 |
| databricks | 2 |
| bigquery | 1 |
| duckdb | 1 |
| flink | 1 |
| hive | 1 |

**26 so far, across 7 dialects**, at 1,300 of 2,262 fixtures. The run is partial — I am posting the corrected shape now rather than sitting on a number I already knew was wrong, and I will post the completed count when the sweep finishes. Every hit so far is `RF06`.

The direction of the correction matters: the problem is **wider** than I reported, not narrower, and it is not a MySQL-family quirk. `bigquery`, `duckdb`, `flink` and `hive` have nothing to do with `'user'@'host'` account specifications, which is what #8462 is about. Whatever RF06 is doing to identifier segments, it is doing it in dialects that share no syntax with the one I originally blamed.

That is more consistent with the root cause in my previous comment — a segment type the grammar at that position cannot accept — than with anything specific to MySQL.

Nothing else in the report changes: the mechanism, the `FunctionNameSegment` / `TypedParser("word")` mismatch, and the two directions for a fix all stand.

**Comment by rubytobi** [comment]
Closing this issue here. 

I understand that sometimes we think we have found a problem but later it turns out to be something different. This issue was opened, then reframed twice. Please review your findings again, reproduce manually, and open a clear issue.
Given you are comfortable using AI, I recommend using a new session with clear context to run an adversarial prompt and verify your finding before opening an issue.

**Comment by WAHIB-EL-KHADIRI** [comment]
Fair, and the process criticism lands.

I filed before I had root-caused it, then corrected the attribution and then the scope. Each correction was right on its own, but three versions of one report reads as thrashing and costs you time either way. That is on me, not on the finding.

I'll do what you asked — verify it manually end to end, and come back with one issue that is correct when it is filed, or not come back at all.

Thanks for saying it plainly rather than just closing it.
